spark/src/main/java/org/opensearch/sql/spark/dispatcher/SparkQueryDispatcher.java

```diff
@@ -39,65 +35,68 @@ public class SparkQueryDispatcher {
+
+    if (LangType.SQL.equals(dispatchQueryRequest.getLangType())
+        && SQLQueryUtils.isFlintExtensionQuery(dispatchQueryRequest.getQuery())) {
+      IndexQueryDetails indexQueryDetails = getIndexQueryDetails(dispatchQueryRequest);
+      DispatchQueryContext context = getDefaultDispatchContextBuilder(dispatchQueryRequest, dataSourceMetadata)
+                      .indexQueryDetails(indexQueryDetails)
+                      .build();
+
+      return getQueryHandlerForFlintExtensionQuery(indexQueryDetails, emrServerlessClient)
+              .submit(dispatchQueryRequest, context);
+    } else {
+      DispatchQueryContext context = getDefaultDispatchContextBuilder(dispatchQueryRequest, dataSourceMetadata)
+                      .build();
+      return getDefaultAsyncQueryHandler(emrServerlessClient).submit(dispatchQueryRequest, context);
+    }
+  }
+
+  private static DispatchQueryContext.DispatchQueryContextBuilder getDefaultDispatchContextBuilder(DispatchQueryRequest dispatchQueryRequest, DataSourceMetadata dataSourceMetadata) {
+    return DispatchQueryContext.builder()
             .dataSourceMetadata(dataSourceMetadata)
             .tags(getDefaultTagsForJobSubmission(dispatchQueryRequest))
             .queryId(AsyncQueryId.newAsyncQueryId(dataSourceMetadata.getName()));
+  }
 
-    // override asyncQueryHandler with specific.
-    if (LangType.SQL.equals(dispatchQueryRequest.getLangType())
-        && SQLQueryUtils.isFlintExtensionQuery(dispatchQueryRequest.getQuery())) {
-      IndexQueryDetails indexQueryDetails =
-          SQLQueryUtils.extractIndexDetails(dispatchQueryRequest.getQuery());
-      fillMissingDetails(dispatchQueryRequest, indexQueryDetails);
-      contextBuilder.indexQueryDetails(indexQueryDetails);
-
-      if (isEligibleForIndexDMLHandling(indexQueryDetails)) {
-        asyncQueryHandler = createIndexDMLHandler(emrServerlessClient);
-      } else if (isEligibleForStreamingQuery(indexQueryDetails)) {
-        asyncQueryHandler =
-            new StreamingQueryHandler(
-                emrServerlessClient, jobExecutionResponseReader, leaseManager);
-      } else if (IndexQueryActionType.REFRESH.equals(indexQueryDetails.getIndexQueryActionType())) {
-        // manual refresh should be handled by batch handler
-        asyncQueryHandler =
-            new RefreshQueryHandler(
-                emrServerlessClient,
-                jobExecutionResponseReader,
-                flintIndexMetadataService,
-                stateStore,
-                leaseManager);
-      }
+  private AsyncQueryHandler getQueryHandlerForFlintExtensionQuery(IndexQueryDetails indexQueryDetails, EMRServerlessClient emrServerlessClient) {
+    if (isEligibleForIndexDMLHandling(indexQueryDetails)) {
+      return queryHandlerFactory.getIndexDMLHandler(emrServerlessClient);
+    } else if (isEligibleForStreamingQuery(indexQueryDetails)) {
+      return queryHandlerFactory.getStreamingQueryHandler(emrServerlessClient);
```

Comment:
Do we need EMRServerlessClient in this class?

Reply:
Good point. Fixed.